pep-9999.rst

```diff
@@ -0,0 +1,257 @@
+PEP: 9999
+Title: Package index configuration standardization
+Author: Fredrik Averpil <fredrik at averpil.com>
+Status: Draft
+Type: Informational
+Created: 05-May-2021
+Post-History: <dates of postings to python-ideas and/or python-dev>
+
+
+Abstract
+========
+
+This PEP standardizes configuration of PyPI repositories.
+
+
+Motivation
+==========
+
+Workflow tools come up with their own configuration mechanisms for PyPI repositories, as there is no standard for this. This causes fragmentation and pain points.
+
+Examples of such configuration files:
+
+* Pip's ``pip.conf`` and ``pip.ini``
+* Packaging's ``.pypirc`` (also used by tools like Flit and Twine)
+* Poetry's ``config.toml`` and ``auth.toml``
+
+
+Rationale
+=========
+
+While this PEP aim to come up with a unified PyPI repository configuration, thought has been given so to be able to extend the configuration later. For example, PyPI repositories are assumed in this PEP, but a ``type`` attribute could be introduced which would default to ``pypi`` down the line, to allow for other package repository types. Any tool could also start taking advantage of the fact that the configuration files and their discovery is a standard, and use this for their own configuration.
+
+
+Specification
+=============
+
+The specification consists of three parts: the configuration definition, package discovery and configuration file discovery.
+
+
+Configuration definition
+------------------------
+
+The configuration is split into two files:
+
+* ``config.toml`` for PyPI repository configuration
+* ``auth.toml`` for credentials
+
+
+The config.toml file
+''''''''''''''''''''
+
+Package index configuration is defined in the ``config.toml`` file::
+
+
+    [repositories]
+
+    [repostiories.pypi]
+    url = "https://pypi.org/simple/"
+
+    [repostiories.myrepo1]
+    url = "http://myserver2.org/simple"
+
+    [authentication]
+
+    [authentication.pypi]
+    username = "__token__"
+    password = "pypi-AgEIcHlwaS5vcmcCJDJjMmI5NDJmLWIzNzQtNDc3ZC04ZmRkLTM4NDEzY2M1ZTU5MAACJXsicGVybWlzc2lvbnMiOiAidXNlciIsICJ2ZXJzaW9uIjogMX0AAAYgV91r8iEUUz7Q5xbsfi_KGiO3wrJR2_IOOpnFyLo62bk"
+
+    [authentication.myrepo1]
+    username = "foo"
+    password = "YmFyCg=="
+
+
+A repository has the following attributes::
+
+    ==================  ===========  =============================================
+    Attribute           Type         Description
+    ==================  ===========  =============================================
+    final               bool         Make it impossible to override/remove the repository configuration (default: false)
```

Comment:
Is this something that has actual value? This can make run time behaviour harder to reason about. Further this can have issues when using PEP 610 based installs. And also causes issues where lock files are relied upon.

Reply:
Yes, but it should only affect the task of merging multiple configuration files. So there'll be a single (shared) function to "load repository configuration" that will use this, and then tools will consume the result.

It's very useful when you want to (e.g.) replace regular PyPI with an internal cache/mirror on a build machine, because you can set a machine-wide config and mark it `final` to avoid project/user/etc. configs from overriding it. (And then you need to prevent your team from using tool-specific configuration, but that's a people problem that we can't fix in this spec.)